## Fama french regression in python

$ r_{i} - r_{f} = \alpha_{i} + \beta_i (r_{M} - r_{f}) + s_i \cdot SMB + h_i \cdot HML + \epsilon_i$


where 

$ r_{i} - r_{f}$ is called excess return 

SMB captures size effect

HML captures value effect


In [2]:
import pandas_datareader.data as reader
import pandas as pd
import datetime as dt
import statsmodels.api as sm
 


In [12]:
import yfinance


end = dt.date(2024, 5, 30)
start = end + dt.timedelta(days = -5 * 365)
# FDGRX fidelity
tickers = ['FDGRX']

data_df = yfinance.download(tickers, start, end)
data_df.head()

[*********************100%%**********************]  1 of 1 completed


Open       High        Low      Close  Adj Close  Volume
Date                                                                     
2019-06-03  18.010000  18.010000  18.010000  18.010000  12.973203       0
2019-06-04  18.500000  18.500000  18.500000  18.500000  13.326166       0
2019-06-05  18.690001  18.690001  18.690001  18.690001  13.463032       0
2019-06-06  18.750000  18.750000  18.750000  18.750000  13.506250       0
2019-06-07  19.059999  19.059999  19.059999  19.059999  13.729552       0

In [13]:
data_df = data_df['Adj Close']
data_df.head()

Date
2019-06-03    12.973203
2019-06-04    13.326166
2019-06-05    13.463032
2019-06-06    13.506250
2019-06-07    13.729552
Name: Adj Close, dtype: float64

In [19]:
daily_returns = data_df.pct_change().dropna(axis = 0)
daily_returns.head()

Date
2019-06-04    0.027207
2019-06-05    0.010270
2019-06-06    0.003210
2019-06-07    0.016533
2019-06-10    0.003673
Name: Adj Close, dtype: float64

In [25]:
monthly_returns = daily_returns\
                    .resample('M')\
                    .agg(lambda x: (x+1).prod() - 1)
monthly_returns.shape

(60,)

In [27]:
monthly_returns.head()

Date
2019-06-30    0.088284
2019-07-31    0.016837
2019-08-31   -0.013046
2019-09-30   -0.016777
2019-10-31    0.038780
Freq: M, Name: Adj Close, dtype: float64

$ r_{i} - r_{f} = \alpha_{i} + \beta_i (r_{M} - r_{f}) + s_i \cdot SMB + h_i \cdot HML + \epsilon_i$

We want to get market factor , the SMB and HML. We will use kenneth french's data library


In [38]:
# BUT we forgot to take rf rate under consideration
# https://mba.tuck.dartmouth.edu/pages/faculty/ken.french/data_library.html
factors = reader.DataReader('F-F_Research_Data_Factors', 
                       'famafrench', 
                       start = monthly_returns.index[0],
                       end = monthly_returns.index[-1]
                       )

/var/folders/mw/pvq5xc9d1klbjmpfrl8cnwfm0000gn/T/ipykernel_13213/3215237425.py:3: FutureWarning: The argument 'date_parser' is deprecated and will be removed in a future version. Please use 'date_format' instead, or read your data in as 'object' dtype and then call 'to_datetime'.
  factors = reader.DataReader('F-F_Research_Data_Factors',
/var/folders/mw/pvq5xc9d1klbjmpfrl8cnwfm0000gn/T/ipykernel_13213/3215237425.py:3: FutureWarning: The argument 'date_parser' is deprecated and will be removed in a future version. Please use 'date_format' instead, or read your data in as 'object' dtype and then call 'to_datetime'.
  factors = reader.DataReader('F-F_Research_Data_Factors',


In [39]:
print(f"start={start} end={end}")
print(factors.keys())
print(factors[0].head())
print(factors[1].head())
print(factors['DESCR'])


start=2019-06-01 end=2024-05-30
dict_keys([0, 1, 'DESCR'])
         Mkt-RF   SMB   HML    RF
Date                             
2019-06    6.93  0.29 -0.71  0.18
2019-07    1.19 -1.93  0.48  0.19
2019-08   -2.58 -2.38 -4.78  0.16
2019-09    1.43 -0.96  6.75  0.18
2019-10    2.06  0.29 -1.91  0.16
      Mkt-RF    SMB    HML    RF
Date                            
2019   28.28  -6.14 -10.46  2.15
2020   23.66  13.18 -46.67  0.45
2021   23.56  -3.89  25.49  0.04
2022  -21.60  -6.95  25.81  1.43
2023   21.70  -3.24 -13.60  4.95
F-F Research Data Factors
-------------------------

This file was created by CMPT_ME_BEME_RETS using the 202404 CRSP database. The 1-month TBill return is from Ibbotson and Associates, Inc. Copyright 2024 Kenneth R. French

  0 : (59 rows x 4 cols)
  1 : Annual Factors: January-December (5 rows x 4 cols)


In [48]:
# we want monthly returns
print(factors[0].head())
print(factors[0].tail())




         Mkt-RF   SMB   HML    RF
Date                             
2019-06    6.93  0.29 -0.71  0.18
2019-07    1.19 -1.93  0.48  0.19
2019-08   -2.58 -2.38 -4.78  0.16
2019-09    1.43 -0.96  6.75  0.18
2019-10    2.06  0.29 -1.91  0.16
         Mkt-RF   SMB   HML    RF
Date                             
2023-12    4.87  6.34  4.93  0.43
2024-01    0.70 -5.09 -2.38  0.47
2024-02    5.06 -0.24 -3.49  0.42
2024-03    2.83 -2.49  4.19  0.43
2024-04   -4.67 -2.39 -0.51  0.47


In [47]:
# Lets we what we have in our df => starts from June. Which matches with above
print(monthly_returns.head())
print(monthly_returns.tail())

Date
2019-06-30    0.088284
2019-07-31    0.016837
2019-08-31   -0.013046
2019-09-30   -0.016777
2019-10-31    0.038780
Freq: M, Name: Adj Close, dtype: float64
Date
2024-01-31    0.030388
2024-02-29    0.090909
2024-03-31    0.026756
2024-04-30   -0.044517
2024-05-31    0.096023
Freq: M, Name: Adj Close, dtype: float64


In [45]:
monthly_returns.shape
factors[0].shape

(59, 4)

In [49]:
# index of monthly_returns is 2019-06-30 format but index of factors[0].tail() is 2019-06 format
# let me try converting 2019-06-30 to 2019-06


In [56]:
monthly_returns.index = monthly_returns.index.to_period()
monthly_returns.head()

Date
2019-06    0.088284
2019-07    0.016837
2019-08   -0.013046
2019-09   -0.016777
2019-10    0.038780
Freq: M, Name: Adj Close, dtype: float64

In [52]:
factors[0].index

PeriodIndex(['2019-06', '2019-07', '2019-08', '2019-09', '2019-10', '2019-11',
             '2019-12', '2020-01', '2020-02', '2020-03', '2020-04', '2020-05',
             '2020-06', '2020-07', '2020-08', '2020-09', '2020-10', '2020-11',
             '2020-12', '2021-01', '2021-02', '2021-03', '2021-04', '2021-05',
             '2021-06', '2021-07', '2021-08', '2021-09', '2021-10', '2021-11',
             '2021-12', '2022-01', '2022-02', '2022-03', '2022-04', '2022-05',
             '2022-06', '2022-07', '2022-08', '2022-09', '2022-10', '2022-11',
             '2022-12', '2023-01', '2023-02', '2023-03', '2023-04', '2023-05',
             '2023-06', '2023-07', '2023-08', '2023-09', '2023-10', '2023-11',
             '2023-12', '2024-01', '2024-02', '2024-03', '2024-04'],
            dtype='period[M]', name='Date')

In [60]:
monthly_returns.to_frame().head()
factors[0]

Mkt-RF   SMB    HML    RF
Date                              
2019-06    6.93  0.29  -0.71  0.18
2019-07    1.19 -1.93   0.48  0.19
2019-08   -2.58 -2.38  -4.78  0.16
2019-09    1.43 -0.96   6.75  0.18
2019-10    2.06  0.29  -1.91  0.16
2019-11    3.87  0.77  -2.02  0.12
2019-12    2.77  0.73   1.75  0.14
2020-01   -0.11 -3.11  -6.25  0.13
2020-02   -8.13  1.07  -3.81  0.12
2020-03  -13.39 -4.83 -13.87  0.13
2020-04   13.65  2.45  -1.33  0.00
2020-05    5.58  2.47  -4.88  0.01
2020-06    2.46  2.69  -2.20  0.01
2020-07    5.77 -2.33  -1.41  0.01
2020-08    7.63 -0.22  -2.97  0.01
2020-09   -3.63  0.02  -2.71  0.01
2020-10   -2.10  4.38   4.25  0.01
2020-11   12.47  5.80   2.09  0.01
2020-12    4.63  4.89  -1.51  0.01
2021-01   -0.03  7.36   3.01  0.01
2021-02    2.78  2.05   7.15  0.00
2021-03    3.08 -2.37   7.39  0.00
2021-04    4.93 -3.19  -0.95  0.00
2021-05    0.29 -0.25   7.09  0.00
2021-06    2.75  1.68  -7.87  0.00
2021-07    1.27 -3.97  -1.78  0.00
2021-08    2.91 -0.42  -0.15  0.00
2021-09   -4.37  0.71   5.08  0.00
2021-10    6.65 -2.35  -0.49  0.00
2021-11   -1.55 -1.32  -0.45  0.00
2021-12    3.10 -1.67   3.26  0.01
2022-01   -6.25 -5.94  12.75  0.00
2022-02   -2.29  2.19   3.08  0.00
2022-03    3.05 -1.61  -1.81  0.01
2022-04   -9.46 -1.39   6.17  0.01
2022-05   -0.34 -1.83   8.39  0.03
2022-06   -8.43  2.10  -5.97  0.06
2022-07    9.57  2.81  -4.05  0.08
2022-08   -3.77  1.40   0.29  0.19
2022-09   -9.35 -0.81   0.05  0.19
2022-10    7.83  0.06   8.01  0.23
2022-11    4.60 -3.52   1.38  0.29
2022-12   -6.41 -0.69   1.37  0.33
2023-01    6.65  5.01  -4.01  0.35
2023-02   -2.58  1.17  -0.81  0.34
2023-03    2.51 -5.51  -8.85  0.36
2023-04    0.61 -3.35  -0.04  0.35
2023-05    0.35  1.61  -7.72  0.36
2023-06    6.46  1.54  -0.26  0.40
2023-07    3.21  2.08   4.11  0.45
2023-08   -2.39 -3.16  -1.06  0.45
2023-09   -5.24 -2.51   1.52  0.43
2023-10   -3.19 -3.88   0.18  0.47
2023-11    8.84 -0.02   1.64  0.44
2023-12    4.87  6.34   4.93  0.43
2024-01    0.70 -5.09  -2.38  0.47
2024-02    5.06 -0.24  -3.49  0.42
2024-03    2.83 -2.49   4.19  0.43
2024-04   -4.67 -2.39  -0.51  0.47

In [63]:
merged_df = monthly_returns.to_frame().join(factors[0])
merged_df.head()

Adj Close  Mkt-RF   SMB   HML    RF
Date                                        
2019-06   0.088284    6.93  0.29 -0.71  0.18
2019-07   0.016837    1.19 -1.93  0.48  0.19
2019-08  -0.013046   -2.58 -2.38 -4.78  0.16
2019-09  -0.016777    1.43 -0.96  6.75  0.18
2019-10   0.038780    2.06  0.29 -1.91  0.16

In [65]:
merged_df = merged_df.dropna(axis=0)

In [66]:
merged_df.tail()

Adj Close  Mkt-RF   SMB   HML    RF
Date                                        
2023-12   0.059996    4.87  6.34  4.93  0.43
2024-01   0.030388    0.70 -5.09 -2.38  0.47
2024-02   0.090909    5.06 -0.24 -3.49  0.42
2024-03   0.026756    2.83 -2.49  4.19  0.43
2024-04  -0.044517   -4.67 -2.39 -0.51  0.47

In [68]:
merged_df[['Mkt-RF','SMB',	'HML', 'RF']] = merged_df[['Mkt-RF','SMB',	'HML', 'RF']]/100


In [69]:
# we have monthly retyrnsn in Adj Close
merged_df['returns-rf'] = merged_df["Adj Close"] - merged_df.RF

In [71]:
merged_df

Adj Close  Mkt-RF     SMB     HML      RF  returns-rf
Date                                                          
2019-06   0.088284  0.0693  0.0029 -0.0071  0.0018    0.086484
2019-07   0.016837  0.0119 -0.0193  0.0048  0.0019    0.014937
2019-08  -0.013046 -0.0258 -0.0238 -0.0478  0.0016   -0.014646
2019-09  -0.016777  0.0143 -0.0096  0.0675  0.0018   -0.018577
2019-10   0.038780  0.0206  0.0029 -0.0191  0.0016    0.037180
2019-11   0.072175  0.0387  0.0077 -0.0202  0.0012    0.070975
2019-12   0.029354  0.0277  0.0073  0.0175  0.0014    0.027954
2020-01   0.027622 -0.0011 -0.0311 -0.0625  0.0013    0.026322
2020-02  -0.044191 -0.0813  0.0107 -0.0381  0.0012   -0.045391
2020-03  -0.102479 -0.1339 -0.0483 -0.1387  0.0013   -0.103779
2020-04   0.183749  0.1365  0.0245 -0.0133  0.0000    0.183749
2020-05   0.101839  0.0558  0.0247 -0.0488  0.0001    0.101739
2020-06   0.077769  0.0246  0.0269 -0.0220  0.0001    0.077669
2020-07   0.081224  0.0577 -0.0233 -0.0141  0.0001    0.081124
2020-08   0.126135  0.0763 -0.0022 -0.0297  0.0001    0.126035
2020-09  -0.033199 -0.0363  0.0002 -0.0271  0.0001   -0.033299
2020-10  -0.028562 -0.0210  0.0438  0.0425  0.0001   -0.028662
2020-11   0.138751  0.1247  0.0580  0.0209  0.0001    0.138651
2020-12   0.037423  0.0463  0.0489 -0.0151  0.0001    0.037323
2021-01   0.013995 -0.0003  0.0736  0.0301  0.0001    0.013895
2021-02   0.020702  0.0278  0.0205  0.0715  0.0000    0.020702
2021-03  -0.010288  0.0308 -0.0237  0.0739  0.0000   -0.010288
2021-04   0.064449  0.0493 -0.0319 -0.0095  0.0000    0.064449
2021-05  -0.011719  0.0029 -0.0025  0.0709  0.0000   -0.011719
2021-06   0.073687  0.0275  0.0168 -0.0787  0.0000    0.073687
2021-07   0.003681  0.0127 -0.0397 -0.0178  0.0000    0.003681
2021-08   0.045062  0.0291 -0.0042 -0.0015  0.0000    0.045062
2021-09  -0.050890 -0.0437  0.0071  0.0508  0.0000   -0.050890
2021-10   0.082673  0.0665 -0.0235 -0.0049  0.0000    0.082673
2021-11   0.017321 -0.0155 -0.0132 -0.0045  0.0000    0.017321
2021-12  -0.033866  0.0310 -0.0167  0.0326  0.0001   -0.033966
2022-01  -0.111873 -0.0625 -0.0594  0.1275  0.0000   -0.111873
2022-02  -0.033344 -0.0229  0.0219  0.0308  0.0000   -0.033344
2022-03   0.037368  0.0305 -0.0161 -0.0181  0.0001    0.037268
2022-04  -0.144397 -0.0946 -0.0139  0.0617  0.0001   -0.144497
2022-05  -0.040662 -0.0034 -0.0183  0.0839  0.0003   -0.040962
2022-06  -0.086647 -0.0843  0.0210 -0.0597  0.0006   -0.087247
2022-07   0.119507  0.0957  0.0281 -0.0405  0.0008    0.118707
2022-08  -0.034850 -0.0377  0.0140  0.0029  0.0019   -0.036750
2022-09  -0.099582 -0.0935 -0.0081  0.0005  0.0019   -0.101482
2022-10   0.065429  0.0783  0.0006  0.0801  0.0023    0.063129
2022-11   0.049128  0.0460 -0.0352  0.0138  0.0029    0.046228
2022-12  -0.089017 -0.0641 -0.0069  0.0137  0.0033   -0.092317
2023-01   0.106175  0.0665  0.0501 -0.0401  0.0035    0.102675
2023-02  -0.012450 -0.0258  0.0117 -0.0081  0.0034   -0.015850
2023-03   0.067914  0.0251 -0.0551 -0.0885  0.0036    0.064314
2023-04   0.009901  0.0061 -0.0335 -0.0004  0.0035    0.006401
2023-05   0.071267  0.0035  0.0161 -0.0772  0.0036    0.067667
2023-06   0.064766  0.0646  0.0154 -0.0026  0.0040    0.060766
2023-07   0.041322  0.0321  0.0208  0.0411  0.0045    0.036822
2023-08  -0.015238 -0.0239 -0.0316 -0.0106  0.0045   -0.019738
2023-09  -0.064475 -0.0524 -0.0251  0.0152  0.0043   -0.068775
2023-10  -0.029979 -0.0319 -0.0388  0.0018  0.0047   -0.034679
2023-11   0.110835  0.0884 -0.0002  0.0164  0.0044    0.106435
2023-12   0.059996  0.0487  0.0634  0.0493  0.0043    0.055696
2024-01   0.030388  0.0070 -0.0509 -0.0238  0.0047    0.025688
2024-02   0.090909  0.0506 -0.0024 -0.0349  0.0042    0.086709
2024-03   0.026756  0.0283 -0.0249  0.0419  0.0043    0.022456
2024-04  -0.044517 -0.0467 -0.0239 -0.0051  0.0047   -0.049217

$ r_{i} - r_{f} = \alpha_{i} + \beta_i (r_{M} - r_{f}) + s_i \cdot SMB + h_i \cdot HML + \epsilon_i$

we can run regression now

In [72]:
## Regression
y = merged_df['returns-rf']
X = merged_df[["Mkt-RF","SMB",	"HML"]]

# intercept not added by default, we need to add
X_sm = sm.add_constant(X)
X_sm.head()


const  Mkt-RF     SMB     HML
Date                                  
2019-06    1.0  0.0693  0.0029 -0.0071
2019-07    1.0  0.0119 -0.0193  0.0048
2019-08    1.0 -0.0258 -0.0238 -0.0478
2019-09    1.0  0.0143 -0.0096  0.0675
2019-10    1.0  0.0206  0.0029 -0.0191

In [73]:
model = sm.OLS(y,X_sm)
results = model.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             returns-rf   R-squared:                       0.953
Model:                            OLS   Adj. R-squared:                  0.951
Method:                 Least Squares   F-statistic:                     372.5
Date:                Mon, 03 Jun 2024   Prob (F-statistic):           1.68e-36
Time:                        16:43:02   Log-Likelihood:                 165.34
No. Observations:                  59   AIC:                            -322.7
Df Residuals:                      55   BIC:                            -314.4
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0056      0.002      2.755      0.008       0.002       0.010
Mkt-RF         1.1470      0.038     29.881      0.000       1.070       1.224
SMB            0.1603      0.071      2.258      0.028       0.018       0.302
HML           -0.4284      0.042    -10.172      0.000      -0.513      -0.344
==============================================================================
Omnibus:                       19.567   Durbin-Watson:                   2.356
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               36.217
Skew:                          -1.041   Prob(JB):                     1.37e-08
Kurtosis:                       6.225   Cond. No.                         36.6
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Models turns out to be 

$ r_{i} - r_{f} = 0.0056 + 1.1470 * (r_{M} - r_{f}) + 0.1603 \cdot SMB -0.4284 \cdot HML + \epsilon_i$


Note that r^2 is quite high 0.953

$s_i = 0.1603$ is positive , this means fund FDGRX is investing in small stocks as compared to big stocks. (excess return is more correlated to s Minus Big factor)

$h_i =  -0.4284$ his means fund FDGRX is investing in growth more than value stocks ( H - L). it is expected as Funds name id Fidelity Growth..


$ alpha = const	 = 0.0056$ which means it is a return above the risk factors which is good.